README.md
# psichomics

<!-- badges: start -->
[![Travis Build Status][travisIcon]][travis]
[![AppVeyor Build Status][appVeyorIcon]][appVeyor]
[![codecov][codecovIcon]][codecov]
<!-- badges: end -->

> **Original article:**
>
> Nuno Saraiva-Agostinho and Nuno L. Barbosa-Morais (2019). 
[psichomics: graphical application for alternative splicing quantification and analysis][article].
*Nucleic Acids Research*. 47(2), e7.

Interactive R package with an intuitive Shiny-based graphical 
interface for alternative splicing quantification and integrative analyses of
alternative splicing and gene expression based on 
[The Cancer Genome Atlas (TCGA)][TCGA], the 
[Genotype-Tissue Expression (GTEx) project][GTEx], 
[Sequence Read Archive (SRA)][SRA] and user-provided data.

*psichomics* interactively performs survival, dimensionality reduction and 
median- and variance-based differential splicing and gene expression analyses
that benefit from the incorporation of clinical and molecular sample-associated
features (such as tumour stage or survival). Interactive visual access to
genomic mapping and functional annotation of selected alternative splicing
events is also included.

![Differential splicing analysis in *psichomics*](man/figures/screenshot.png)

## Table of Contents

* [Install and start running](#install-and-start-running)
    * [Bioconductor release](#bioconductor-release)
    * [GitHub version](#github-version)
* [Tutorials](#tutorials)
* [Workflow](#workflow)
    * [Data input](#data-input)
        * [Splicing quantification](#splicing-quantification)
        * [Gene expression processing](#gene-expression-processing)
    * [Data grouping](#data-grouping)
    * [Data analyses](#data-analyses)
* [Feedback and support](#feedback-and-support)
* [References](#references)

## Install and start running

To install the package from [Bioconductor][bioconductor], type the following in 
[RStudio][rstudio] or in an R console:

```r
install.packages("BiocManager")
BiocManager::install("psichomics")
```

## Tutorials

The following case studies and tutorials are available and were based on our 
[original article][article]:

* [Visual interface][tutorial-gui]
* [Command-line interface][tutorial-cli]
* [Loading SRA and user-provided RNA-seq data][tutorial-custom-data]
* [Preparing alternative splicing annotations][tutorial-prep-AS-annotation]

## Workflow

### Data input

Automatic retrieval and loading of pre-processed data from the following sources:

* [TCGA][TCGA] data of given tumours, including subject- and sample-associated
information, junction quantification and gene expression data
* [GTEx][GTEx] data of given tissues, including subject- and sample-associated
information, junction quantification and gene expression data
* [SRA][SRA] data from select SRA projects via the [recount2][recount2] package

Other SRA and user-provided data can be manually aligned and loaded. Please read
[Loading SRA and user-provided RNA-seq data][tutorial-custom-data].

#### Splicing quantification

The quantification of each alternative splicing event is based on the proportion
of junction reads that support the inclusion isoform, known as percent 
spliced-in or PSI [(Wang *et al.*, 2008)][Wang2008].

An estimate of this value is obtained based on the the proportion of reads 
supporting the inclusion of an exon over the reads supporting both the inclusion
and exclusion of that exon. To measure this estimate, we require:

1. Alternative splicing annotation: human (hg19 and hg38 assemblies) annotation
is provided and custom annotations can be used.
2. Quantification of RNA-Seq reads aligning to splice junctions (junction
quantification), either user-provided or retrieved from [TCGA][TCGA],
[GTEx][GTEx] and [SRA][SRA].

#### Gene expression processing

Gene expression can be normalised, filtered and log2-transformed in-app or
provided by the user.

### Data grouping

Molecular and clinical sample-associated attributes allow to establish groups 
that can be explored in data analyses.

For instance, [TCGA][TCGA] data can be analysed based on smoking history, gender
and race, among other attributes. Groups can also be manipulated (e.g. merged,
intersected, etc.), allowing for complex attribute combinations. Groups can also
be saved and loaded between sessions.

### Data Analyses

* **Dimensionality reduction** via principal and independent component analysis
(PCA and ICA) on alternative splicing quantification and gene expression.

* **Differential splicing and gene expression analysis** based on variance and
median parametric and non-parametric statistical tests.

* **Correlation between gene expression and splicing quantification**, useful to
correlate the expression of a given event with the expression of RNA-binding
proteins, for instance.

* **Survival analysis** via Kaplan-Meier curves and Cox models based on
sample-associated features. Additionally, we can study the impact of a splicing
event (based on its quantification) or a gene (based on its gene expression) on
patient survivability.

* **Gene, transcript and protein annotation** including relevant research
articles

## Feedback and support

Please send any feedback and questions on psichomics to:

> Nuno Saraiva-Agostinho ([nunoagostinho@medicina.ulisboa.pt](mailto:nunoagostinho@medicina.ulisboa.pt))
> 
> [Disease Transcriptomics Lab, Instituto de Medicina Molecular (Portugal)][NMorais]

## References

Wang, E. T., R. Sandberg, S. Luo, I. Khrebtukova, L. Zhang, C. Mayr, S. F. 
Kingsmore, G. P. Schroth, and C. B. Burge. 2008. 
[*Alternative isoform regulation in human tissue transcriptomes.*][Wang2008] 
Nature 456 (7221): 470–76.

[TCGA]: https://tcga-data.nci.nih.gov
[bioconductor]: https://www.bioconductor.org
[r]: https://www.r-project.org
[rstudio]: https://www.rstudio.com/products/rstudio
[NMorais]: http://imm.medicina.ulisboa.pt/group/distrans/
[conduct]: CONDUCT.md
[Wang2008]: http://www.nature.com/nature/journal/v456/n7221/full/nature07509.html
[travisIcon]: https://travis-ci.org/nuno-agostinho/psichomics.svg?branch=master
[travis]: https://travis-ci.org/nuno-agostinho/psichomics
[appVeyorIcon]: https://ci.appveyor.com/api/projects/status/github/nuno-agostinho/psichomics?branch=master&svg=true
[appVeyor]: https://ci.appveyor.com/project/nuno-agostinho/psichomics
[codecovIcon]: https://codecov.io/gh/nuno-agostinho/psichomics/branch/master/graph/badge.svg
[codecov]: https://codecov.io/gh/nuno-agostinho/psichomics
[GTEx]: http://www.gtexportal.org
[article]: https://doi.org/10.1093/nar/gky888
[SRA]: https://www.ncbi.nlm.nih.gov/sra
[tutorial-gui]: http://rpubs.com/nuno-agostinho/psichomics-tutorial-visual
[tutorial-cli]: http://rpubs.com/nuno-agostinho/psichomics-cli-tutorial
[tutorial-custom-data]: http://rpubs.com/nuno-agostinho/psichomics-custom-data
[tutorial-prep-AS-annotation]: http://rpubs.com/nuno-agostinho/preparing-AS-annotation
[recount2]: https://jhubiostatistics.shinyapps.io/recount/


NEWS
# 1.12.1 (17 January, 2019)

* Fix loading only a portion of TCGA data (with no warning) if too many cohorts
are selected
* User-provided junction quantification loading:
    - Support junction coordinates from mitochondrial, Z and W chromosomes
    - Fix issues with files containing splice junctions within random, alt and
    unknown chromosomes by discarding those rows (a warning is raised)
* Alternative splicing events can now be represented via diagrams:
    - Redesign of alternative splicing event selection (graphical interface)
    - `plotSplicingEvent()` plots diagram representation of alternative
    splicing events
    - In the visual interface, alternative splicing event diagrams were added
    below distribution plots (to quickly illustrate higher and lower values of
    alternative splicing quantification) and in annotation page
* Improve import/export of data groups from/to a file, including colour support
* Tutorials:
    - Copy-edit command-line interface (CLI) tutorial
    - Copy-edit tutorial on custom alternative splicing annotation preparation

## More bug fixes and minor changes

* Data loading and manipulation:
    - Copy-edit information on the format of user-provided files
    - Warn when discarding rows with duplicated rownames after loading
    user-provided files
    - `getGtexDataTypes()` is now exported, as expected
    - `loadFirebrowseData()` now returns expected data when asking for multiple
    datasets (such as in the case of performing a pan-cancer analysis)
    - `parseSplicingEvent()` now returns the coordinates as numeric if 
    `coords = TRUE` and `char = FALSE`
    - Improve dialog when trying to load a local folder without any supported 
    files available
    - Improve tutorial on loading user-provided data
* Alternative splicing annotation:
    - Confirmation dialog is not displayed anymore when creating folder
    (specially useful while running the visual interface)
    - Allow to select cache directory of AnnotationHub (command-line interface)
    - Fix `prepareGeneQuant()` discarding the argument `strandedness` if
    either `stranded` or `stranded (reverse)`
* Data grouping:
    - Samples not associated with any subject are now kept when exporting groups
    to a file
    - Fix issues related with importing groups based on a file with groups of
    splicing events and/or genes
    - Inform user when groups are successfully loaded from a file and whether
    any group elements are discarded in the process
    - Show number of genes contained in pre-made gene lists
    - Discard unavailable genes when creating group of genes based on pre-made
    gene lists (unless these are automatically created at startup)
* Density plot (`plotDistribution()`):
    - Fix visual bug when plotting a group with only one sample (if one data
    point is available for a group, only the rug plot is drawn)
    - After hiding all plot series, rug plots of the different groups can be 
    distinguished based on the Y axis (different arbitrary Y values are given to
    each rug plot series)
    - Rug plot labels now show data values if sample names are not provided
    - Rug plot labels can now be rotated (rotation is not enabled by default
    given Highcharts issues that may occur at different zoom levels and
    depending on proximity between different sample values)
    - Fix misguiding example in function documentation
* Survival analysis (p-value plot):
    - Fix alternative splicing quantification cutoff being selected based on the
    one whose difference has the highest (instead of the lowest) p-value
    - Fix plot line label presenting "p < 0.05" independently of the threshold 
    used for significance
* Annotation:
    - Avoid app crash when searching for PubMed articles too many times
* Improve warning/error alerts:
    - Fix alerts crashing the visual interface
    - Improve message formatting regarding HTML-containing alerts
* Improve unit tests and function documentation

# 1.10.2 (4 Oct, 2019)

Fix warnings in Windows-specific CI tools and Bioconductor

# 1.10.1 (19 Sep, 2019)

Avoid timeouts in CI tools and Bioconductor builds by skipping a specific unit
test

# 1.8.3 (21 April, 2019)

Replace deprecated `R.utils::evalWithTimeout()` with `R.utils::withTimeout()`

# 1.8.2 (26 March, 2019)

* Data loading:
    - GTEx data can now be automatically downloaded and loaded on-demand
    - By default, data table now only displays at most the first 100 columns 
    for performance reasons
* Alternative splicing quantification:
    - Allow to discard samples before alternative splicing quantification
    - In alternative splicing quantification dataset summary, plot
    quantification based on median, variance and range per splicing event across
    samples to provide tools to filter quantification (`plotRowStats()`)
* Gene expression filtering and normalisation:
    - Allow to discard samples before filtering and normalisation
    - Filter low read counts using `edgeR::filterByExpr()`
    - Allow to perform `limma::voom()` on gene expression data (without design
    matrix) and to apply its normalisation methods
    - In gene expression dataset summary, plot distribution of gene expression 
    per sample, distribution of library sizes and gene-wise mean and variance of
    gene expression across samples to provide the user tools to assess gene 
    expression normalisation (`plotGeneExprPerSample()`, `plotDistribution()` 
    and  `plotRowStats()`, respectively)
    - Convert between different gene identifiers (the original identifier is 
    kept in some conditions, read `convertGeneIdentifiers()`); in the visual
    interface, when filtering and normalising gene expression, ENSEMBL 
    identifiers are converted to gene symbols, by default
* Groups:
    - By default, load pre-made lists of genes when loading gene expression or
    loading/performing alternative splicing quantification

R/app.R
#' @importFrom Rcpp sourceCpp
#' @useDynLib psichomics, .registration=TRUE
#' @include globalAccess.R
NULL

# TODO(NunoA): increase allowed size and warn the user to wait for large files
# Refuse files with size greater than the specified
MB = 20 # File size in GB
options(shiny.maxRequestSize = MB * 1024^5)

# Sanitize errors
options(shiny.sanitize.errors = TRUE)

#' psichomics article's link interface
#'
#' @importFrom shiny tags icon
#' 
#' @return HTML elements
#' @keywords internal
linkToArticle <- function() {
    authors <- c("Nuno Saraiva-Agostinho", "Nuno L Barbosa-Morais")
    title   <- paste("psichomics: graphical application for alternative",
                     "splicing quantification and analysis.")
    year    <- 2019
    journal <- "Nucleic Acids Research"
    volume  <- 47
    number  <- 2
    pages   <- "e7"
    
    tags$a(
        target="_blank", href="https://doi.org/10.1093/nar/gky888",
        tags$div(
            class="alert alert-info", role="alert",
            icon("paper-plane-o"), 
            sprintf("%s (%s).", paste(authors, collapse=" and "), year),
            tags$b(title), tags$i(paste0(journal, ".")),
            sprintf("%s(%s), %s", volume, number, pages)))
}

#' Check if a given function should be loaded by the calling module
#' @param loader Character: name of the file responsible to load such function 
#' @param FUN Function
#' @return Boolean vector
#' @keywords internal
loadBy <- function(loader, FUN) {
    attribute <- attr(FUN, "loader")
    if (is.null(attribute))
        return(FALSE)
    else
        return(attribute == loader)
}

#' Matches server functions from a given loader
#' @param ... Extra arguments to pass to server functions
#' @inheritParams getUiFunctions
#' 
#' @importFrom shiny callModule
#' @return Invisible TRUE
#' @keywords internal
getServerFunctions <- function(loader, ..., priority=NULL) {
    # Get all functions ending with "Server"
    server <- ls(getNamespace("psichomics"), all.names=TRUE, pattern="Server$")
    server <- c(priority, server[!server %in% priority])
    
    lapply(server, function(name) {
        # Parse function name to get the function itself
        FUN <- eval(parse(text=name))
        # Check if module should be loaded by app
        if (loadBy(loader, FUN)) {
            # Remove last "Server" from the name and use it as ID
            id <- gsub("Server$", "", name)
            callModule(FUN, id, ...)
        }
    })
    return(invisible(TRUE))
}

#' Matches user interface (UI) functions from a given loader
#' 
#' @param ns Shiny function to create IDs within a namespace
#' @param loader Character: loader to run the functions
#' @param ... Extra arguments to pass to the user interface (UI) functions
#' @param priority Character: name of functions to prioritise by the given
#' order; for instance, \code{c("data", "analyses")} would load \code{data},
#' then \code{analyses} and finally the remaining functions
#' 
#' @return List of functions related to the given loader
#' @keywords internal
getUiFunctions <- function(ns, loader, ..., priority=NULL) {
    # Get all functions ending with "UI"
    ui <- ls(getNamespace("psichomics"), all.names=TRUE, pattern="UI$")
    ui <- c(priority, ui[!ui %in% priority])
    
    # Get the interface of each tab
    uiList <- lapply(ui, function(name) {
        # Parse function name to get the function itself
        FUN <- eval(parse(text=name))
        # Check if module should be loaded by app
        if (loadBy(loader, FUN)) {
            # Remove last "UI" from the name and use it as ID
            id  <- gsub("UI$", "", name)
            res <- FUN(ns(id), ...)
            # Pass all attributes and add identifier
            attributes(res) <- c(attributes(res), attributes(FUN)[-1])
            return(res)
        }
    })
    # Remove NULL elements from list
    uiList <- Filter(Negate(is.null), uiList)
    return(uiList)
}

#' Create a \code{selectize} input available from any page
#' 
#' @param id Character: input identifier
#' @param placeholder Character: input placeholder
#' @param ASevent Boolean: select alternative splicing events?
#' 
#' @importFrom shiny selectizeInput tagAppendAttributes
#' 
#' @return HTML element for a global \code{selectize} input
#' @keywords internal
globalSelectize <- function(id, placeholder, ASevent=FALSE) {
    elem <- paste0(id, "Elem")
    hideElem <- sprintf("$('#%s')[0].style.display = 'none';", id)
    
    onItemAdd <- I(paste0("function(value, $item) {", hideElem, "}"))
    onBlur    <- I(paste0("function() {", hideElem, "}"))
    onType    <- I(paste0(
        "function(value) { $('#selectizeEvent .selectize-dropdown-content')",
        ".unmark().mark(value, {exclude: ['text']}); }"))
    
    render <- NULL
    if (ASevent) render <- I("{ option: renderEvent }")
    
    opts <- list(onItemAdd=onItemAdd, onBlur=onBlur, maxOptions=20,
                 placeholder=placeholder, render=render, highlight=FALSE,
                 onType=onType)
    
    select <- selectizeInput(elem, "", choices=NULL, width="95%", options=opts)
    select[[3]][[1]] <- NULL
    select <- tagAppendAttributes(select, id=id, style=paste(
        "display: none;",
        "position: absolute;",  "margin-top: 5px !important;"))
    return(select)
}

#' Create a special \code{selectize} input in the navigation bar
#' 
#' @inheritParams globalSelectize
#' @param label Character: input label
#' 
#' @return HTML element to be included in a navigation bar
#' @keywords internal
navSelectize <- function(id, label, placeholder=label, ASevent=FALSE) {
    value <- paste0(id, "Value")
    tags$li( tags$div(
        class="navbar-text",
        style="margin-top: 5px !important; margin-bottom: 0px !important;", 
        globalSelectize(id, placeholder, ASevent=ASevent),
        tags$small(tags$b(label), tags$a(
            "Change...", onclick=paste0(
                '$("#', id, '")[0].style.display = "block";',
                '$("#', id, ' > div > select")[0].selectize.clear();',
                '$("#', id, ' > div > select")[0].selectize.focus();'))), 
        tags$br(), uiOutput(value)))
}

#' Modified \code{tabPanel} function to show icon and title
#' 
#' @note Icon is hidden at small viewports
#' 
#' @param title Character: title of the tab
#' @param icon Character: name of the icon
#' @param ... HTML elements to render
#' @param menu Boolean: create a dropdown menu-like tab?
#' 
#' @importFrom shiny navbarMenu tabPanel
#' 
#' @return HTML interface
#' @keywords internal
modTabPanel <- function(title, ..., icon=NULL, menu=FALSE) {
    if (is.null(icon))
        display <- title
    else
        display <- tagList(icon(class="hidden-sm", icon), title)
    
    if (menu)
        navbarMenu(display, ...)
    else
        tabPanel(display, ..., value=title)
}

#' User interface
#' 
#' The user interface (UI) controls the layout and appearance of the app. All
#' CSS modifications are in the file \code{shiny/www/styles.css}
#' 
#' @importFrom shinyjs useShinyjs
#' @importFrom shiny includeCSS includeScript conditionalPanel div h4 icon
#' shinyUI navbarPage tagAppendChild tagAppendAttributes
#' 
#' @return HTML elements
appUI <- function() {
    uiList <- getUiFunctions(paste, "app", modTabPanel,
                             priority=c("dataUI", "analysesUI"))
    
    header <- tagList(
        # Include CSS files
        includeCSS(insideFile("shiny", "www", "animate.min.css")),
        includeCSS(insideFile("shiny", "www", "psichomics.css")),
        # Include JavaScript files
        includeScript(insideFile("shiny", "www", "jquery.mark.min.js")),
        includeScript(insideFile("shiny", "www", "highcharts.ext.js")),
        includeScript(insideFile("shiny", "www", "fuzzy.min.js")),
        includeScript(insideFile("shiny", "www", "jquery.textcomplete.min.js")),
        includeScript(insideFile("shiny", "www", "psichomics.js")),
        conditionalPanel(
            condition="$('html').hasClass('shiny-busy')",
            div(class="text-right", id="loadmessage",
                h4(tags$span(class="label", class="label-info",
                             icon("flask", "fa-spin"), "Working...")))))
    
    nav <- do.call(navbarPage, c(
        list(title="psichomics", id="nav", collapsible=TRUE, 
             header=header, position="fixed-top", footer=useShinyjs()),
        uiList))

    # Hide the header from the navigation bar if the viewport is small
    nav[[3]][[1]][[3]][[1]][[3]][[1]] <- tagAppendAttributes(
        nav[[3]][[1]][[3]][[1]][[3]][[1]], class="hidden-sm")
    
    # Add global selectize input elements to navigation bar
    nav[[3]][[1]][[3]][[1]][[3]][[2]] <- tagAppendChild(
        nav[[3]][[1]][[3]][[1]][[3]][[2]], 
        tags$ul(class="nav navbar-nav navbar-right",
                navSelectize("selectizeCategory", "Selected dataset",
                             "Select dataset"),
                navSelectize("selectizeEvent", "Selected splicing event",
                             "Search by gene and coordinates...",
                             ASevent=TRUE)))
    shinyUI(nav)
}

#' Enable history navigation
#' 
#' Navigate app according to the location given by the navigation bar. Code
#' and logic adapted from
#' \url{https://github.com/daattali/advanced-shiny/blob/master/navigate-history}
#' 
#' @param navId Character: identifier of the navigation bar
#' @param input Input object
#' @param session Session object
#' 
#' @importFrom shiny observe parseQueryString updateTabsetPanel
#' 
#' @inherit psichomics return
#' @keywords internal
browserHistory <- function(navId, input, session) {
    # Update browser history when user changes the active tab
    observeEvent(input[[navId]], {
        autoNav <- getAutoNavigation()
        if (isTRUE(autoNav)) {
            setAutoNavigation(FALSE)
        } else {
            # Update browser history
            runjs(paste0("updateHistory({ page: '", input[[navId]], "'})"))
        }
    })
    
    # Navigate to a tab according to a given query string
    restorePage <- function(qs) {
        data <- parseQueryString(qs)
        if (!is.null(data$page)) {
            setAutoNavigation(TRUE)
            updateTabsetPanel(session, navId, data$page)
        }
    }
    
    # Navigate tabs while browsing history
    observeEvent(input$appLocation, { restorePage(input$appLocation) })
}

#' Server logic
#' 
#' Instructions to build the Shiny app
#' 
#' @param input Shiny input
#' @param output Shiny output
#' @param session Shiny session
#' 
#' @importFrom shiny observe stopApp
#' 
#' @inherit psichomics return
appServer <- function(input, output, session) {
    ns <- session$ns
    groupsServerOnce(input, output, session)
    getServerFunctions("app", priority=c("dataServer", "analysesServer"))
    browserHistory("nav", input, session)
    
    updateSelectizeChoices <- function(session, id, choices, server=FALSE) {
        if (!is.null(choices)) {
            selected <- choices[[1]]
        } else {
            choices  <- list()
            selected <- list()
        }
        updateSelectizeInput(session, id, choices=choices, selected=selected,
                             server=server)
    }
    
    # Update available categories
    observe(updateSelectizeChoices(session, "selectizeCategoryElem", 
                                   names(getData()), server=FALSE))
    
    # Set data category
    observeEvent(input$selectizeCategoryElem, {
        selected <- input$selectizeCategoryElem
        if (!is.null(selected) && selected != "") setCategory(selected)
    })
    
    # Prepare representation of alternative splicing events
    prepareASeventsRepresentation <- reactive({
        ASevent <- getASevents()
        if (!is.null(ASevent)) {
            diagram        <- plotSplicingEvent(ASevent, raw=TRUE,
                                                class="pull-right")
            representation <- ASevent
            names(representation) <- paste(ASevent, "__", diagram)
            
            # Fully extend acronyms for alternative splicing event types
            eventTypes <- getSplicingEventTypes(acronymsAsNames=TRUE)
            for (type in names(eventTypes)) {
                find <- paste0("^", type)
                rep  <- sprintf("%s (%s)", eventTypes[[type]], type)
                names(representation) <- gsub(find, rep, names(representation))
            }
            # Improve chromosome and strand representation
            names(representation) <- gsub("_([0-9A-Za-z].*)_([+-])_",
                                          "_(chr\\1, \\2 strand)_",
                                          names(representation))
        } else {
            representation <- NULL
        }
        return(representation)
    })
    
    # Update available alternative splicing events
    observe({
        representation <- prepareASeventsRepresentation()
        selected <- getASevent()
        if (!is.null(representation) && !is.null(selected)) {
            # Move the selected alternative splicing event to the top
            find <- match(selected, representation)
            sort <- unique(c(find, seq(representation)))
            representation <- representation[sort]
        }   
        updateSelectizeChoices(session, "selectizeEventElem", representation,
                               server=TRUE)
    })
    
    # Set alternative splicing event
    observeEvent(input[["selectizeEventElem"]], {
        selected <- input[["selectizeEventElem"]]
        if (!is.null(selected) && selected != "") setEvent(selected)
    })
    
    # Display selected category
    output$selectizeCategoryValue <- renderUI({
        category <- getCategory()
        if (is.null(category))
            return("No dataset loaded")
        else if(category == "")
            return("No dataset selected")
        else
            return(category)
    })
    
    # Display selected event
    output$selectizeEventValue <- renderUI({
        areEventsLoaded  <- !is.null(getASevents())
        
        selected <- getASevent()
        isSelectionValid <- !is.null(selected) && selected != ""
        
        if (!areEventsLoaded)
            return("No events quantified")
        else if (!isSelectionValid)
            return("No event is selected")
        else
            return(parseSplicingEvent(selected, char=TRUE))
    })
    
    session$onSessionEnded(function() {
        # Stop app and print message to console
        message("\n-- psichomics was closed --")
        suppressMessages(stopApp())
    })
}

#' Start graphical interface of psichomics
#'
#' @inheritParams shiny::runApp
#' @inheritDotParams shiny::runApp -appDir -launch.browser
#' @param reset Boolean: reset Shiny session? Requires package \code{devtools}
#' @param testData Boolean: load with test data
#'
#' @importFrom shiny shinyApp runApp addResourcePath
#'
#' @return \code{NULL} (function is only used to modify the Shiny session's
#' state or internal variables)
#' @export
#'
#' @examples
#' \dontrun{
#' psichomics()
#' }
psichomics <- function(..., launch.browser=TRUE, reset=FALSE, testData=FALSE) {
    # Add icons related to set operations
    addResourcePath("set-operations",
                    insideFile("shiny", "www", "set-operations"))
    if (reset) devtools::load_all()
    
    if (testData) {
        loadFile <- function(file) {
            if (!file.exists(file)) {
                # Fetch file online if not locally available
                link <- paste0("https://github.com/",
                               "nuno-agostinho/psichomics/raw/master/",
                               file)
                file <- url(link)
            }
            readRDS(file)
        }
        data <- NULL
        data[["Clinical data"]]    <- loadFile("vignettes/BRCA_clinical.RDS")
        data[["Gene expression"]]  <- loadFile("vignettes/BRCA_geneExpr.RDS")
        data[["Inclusion levels"]] <- loadFile("vignettes/BRCA_psi.RDS")
        data[["Sample metadata"]]  <- parseTCGAsampleInfo(colnames(
            data[["Inclusion levels"]]))
        setData(list("Test data"=data))
    }
    app <- shinyApp(appUI(), appServer)
    runApp(app, launch.browser=launch.browser, ...)
}